### PR TITLE
feat: add suspense loading files to routes

### PR DESCRIPTION
**Pull request by Nkemjikanma**
## Description

This PR takes care of [issue-1023](https://github.com/typehero/typehero/issues/1023) and creates loading files for routes and creates components that are passed to suspense that load data async

## Related Issue

Closes https://github.com/typehero/typehero/issues/1023

## Screenshots/Video (if applicable):
Screenshot 2023-11-18 at 00 44 27

### CODE REVIEW

File: apps/web/src/app/[locale]/settings/page.tsx [hunk 1]
```diff
@@ -32,7 +34,9 @@ export default async function SettingsPage() {
 
   return (
     <>
-      <Settings user={profileData} />
+      <Suspense fallback={<SettingsSkeleton />}>
```
Comment (aadamw) [inline apps/web/src/app/[locale]/settings/page.tsx:37]: This suspense boundary will not be triggered, unless we move the profileData fetch there

File: apps/web/src/app/[locale]/tracks/_components/track-details-skeleton.tsx [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+          <Skeleton className="h-4 w-96 max-w-[69ch] rounded-lg bg-zinc-300 dark:bg-zinc-700" />
+        </div>
+        <div className="w-24">
+          <Skeleton className="h-12 w-32 rounded-3xl bg-zinc-300 dark:bg-zinc-700" />
+        </div>
+      </div>
+
+      {/* Track Challenges */}
+      <div className="mt-8 flex flex-col gap-2">
+        <TrackChallengesSkeleton />
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/track-details-skeleton.tsx:22]: nit: 
```suggestion
        {new Array(11).fill(true).map((_, idx) => <TrackChallengesSkeleton key={idx} />)}
```

File: apps/web/src/app/[locale]/tracks/_components/track-enrolled-section-carousel.tsx [hunk 3]
```diff
@@ -12,7 +14,9 @@ export function EnrolledTrackSectionCarousel({ tracks }: { tracks: EnrolledTrack
           href={`/tracks/${t.slug}`}
           key={t.id}
         >
-          <PersonalTrackCard track={t} />
+          <Suspense fallback={<EnrolledTrackCardSkeleton />}>
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/track-enrolled-section-carousel.tsx:17]: This suspense boundary will not be triggered unless we fetch track in it

File: apps/web/src/app/[locale]/(profile)/[username]/_components/profile-skeleton.tsx [hunk 4]
```diff
@@ -0,0 +1,40 @@
+import { Skeleton } from '@repo/ui/components/skeleton';
+
+export const ProfileSkeleton = () => {
```
Comment (aadamw) [inline apps/web/src/app/[locale]/(profile)/[username]/_components/profile-skeleton.tsx:3]: suggestion: Could we stick to arrow functions or regular functions?

File: apps/web/src/app/[locale]/settings/_components/settings-skeleton.tsx [hunk 5]
```diff
@@ -0,0 +1,33 @@
+import { Skeleton } from '@repo/ui/components/skeleton';
+
+export const SettingsSkeleton = () => {
```
Comment (aadamw) [inline apps/web/src/app/[locale]/settings/_components/settings-skeleton.tsx:3]: suggestion: Could we stick to arrow functions or regular functions?

File: apps/web/src/app/[locale]/tracks/_components/index.tsx [hunk 6]
```diff
@@ -18,6 +20,7 @@ export async function Tracks() {
     <>
       <div className="flex flex-col gap-5 pb-8 md:gap-10 md:py-5">
         <EnrolledTrackSection />
+
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/index.tsx:23]: nit: not needed line

**Review by aadamw — CHANGES_REQUESTED** [review]

**Comment by aadamw** [comment]
Looks awesome, @Nkemjikanma. I left few comments.

File: apps/web/src/app/[locale]/tracks/_components/track-details-skeleton.tsx (hunk 2 above)
Comment (bautistaaa) [reply apps/web/src/app/[locale]/tracks/_components/track-details-skeleton.tsx:22]:
```ts
  return Array.from({ length: 11 }, (_, i) => (<TrackChallengesSkeleton key={idx} />);
```

is neat too

File: apps/web/src/app/[locale]/(profile)/[username]/_components/profile-skeleton.tsx (hunk 4 above)
Comment (bautistaaa) [reply apps/web/src/app/[locale]/(profile)/[username]/_components/profile-skeleton.tsx:3]: lets use function declaration for component definitions

File: apps/web/src/app/[locale]/settings/page.tsx (hunk 1 above)
Comment (Nkemjikanma) [reply apps/web/src/app/[locale]/settings/page.tsx:37]: Hey man, can you please explain this better? 
I have tested it by delaying the rendering of the component, and it is triggered.

Comment (bautistaaa) [reply apps/web/src/app/[locale]/settings/page.tsx:37]: i’ll pull your branch and verify/look deeper at the code shortly

Comment (aadamw) [reply apps/web/src/app/[locale]/settings/page.tsx]: Wasn't the `loading.tsx` file used in that case? Was the entire page a skeleton or just this section?

File: apps/web/src/app/[locale]/settings/page.tsx [hunk 7]
```diff
@@ -3,6 +3,9 @@ import { prisma } from '@repo/db';
 import type { Metadata } from 'next';
 import { Settings } from './_components/settings';
 import { buildMetaForDefault } from '~/app/metadata';
+import { Suspense } from 'react';
+import SettingsSkeleton from './loading';
+import { de } from 'date-fns/locale';
```
Comment (aadamw) [inline apps/web/src/app/[locale]/settings/page.tsx:8]: Those 3 doesn't seem to be used

File: apps/web/src/app/[locale]/tracks/_components/enrolled-tracks-skeleton.tsx [hunk 8]
```diff
@@ -0,0 +1,49 @@
+import {
+  Card,
+  CardContent,
+  CardDescription,
+  CardHeader,
+  CardTitle,
+} from '@repo/ui/components/card';
+import { Skeleton } from '@repo/ui/components/skeleton';
+import En from '~/locales/en';
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/enrolled-tracks-skeleton.tsx:9]: `En` not used. @bautistaaa maybe we can add eslint rule for unused imports?

File: apps/web/src/app/[locale]/settings/page.tsx (hunk 1 above)
Comment (Nkemjikanma) [reply apps/web/src/app/[locale]/settings/page.tsx]: The `loading.tsx` file was used. 
I also removed the skeleton in that section in my most recent push. 

I think i get what you mean now, so i removed the suspense for just that section. It now relies solely on `loading.tsx`

File: apps/web/src/app/[locale]/tracks/_components/enrolled-tracks-skeleton.tsx (hunk 8 above)
Comment (bautistaaa) [reply apps/web/src/app/[locale]/tracks/_components/enrolled-tracks-skeleton.tsx:9]: ill make a GH issue

Comment (bautistaaa) [reply apps/web/src/app/[locale]/tracks/_components/enrolled-tracks-skeleton.tsx:9]: ill make a GH issue

File: apps/web/src/app/[locale]/(profile)/[username]/layout.tsx [hunk 9]
```diff
@@ -7,5 +9,9 @@ export default async function LayoutProfile({
   children: React.ReactNode;
   params: { username: string };
 }) {
-  return <Profile username={username}>{children}</Profile>;
+  return (
+    <Suspense fallback={<ProfileSkeleton />}>
```
Comment (bautistaaa) [inline apps/web/src/app/[locale]/(profile)/[username]/layout.tsx:13]: can we use `loading.tsx` here?

Comment (Nkemjikanma) [reply apps/web/src/app/[locale]/(profile)/[username]/layout.tsx]: Done ser! 🚀

File: apps/web/src/app/[locale]/settings/_components/edit-profile-tab.tsx [hunk 10]
```diff
@@ -14,6 +14,8 @@ import { RichMarkdownEditor } from '~/components/rich-markdown-editor';
 
 import { updateProfile } from './settings.action';
 import { profileSchema, type ProfileSchema } from './schema';
+import { SettingsSkeleton } from './settings-skeleton';
+import { Suspense } from 'react';
```
Comment (aadamw) [inline apps/web/src/app/[locale]/settings/_components/edit-profile-tab.tsx:18]: Not used

File: apps/web/src/app/[locale]/tracks/_components/track-enrolled-section.tsx [hunk 11]
```diff
@@ -2,6 +2,8 @@ import { auth, type Session } from '@repo/auth/server';
 import { prisma } from '@repo/db';
 import { EnrolledTrackSectionCarousel } from './track-enrolled-section-carousel';
 import { SkeletonTrack } from './SkeletonTrack';
+import { Suspense } from 'react';
+import { EnrolledTracksSkeleton } from './enrolled-tracks-skeleton';
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/track-enrolled-section.tsx:6]: Not used

File: apps/web/src/app/[locale]/tracks/_components/track-enrolled-section-carousel.tsx [hunk 12]
```diff
@@ -2,6 +2,8 @@ import Link from 'next/link';
 import { Carousel } from '~/components/carousel';
 import { PersonalTrackCard } from './personal-track-card';
 import type { EnrolledTracks } from './track.action';
+import { EnrolledTrackCardSkeleton } from './enrolled-track-card-skeleton';
+import { Suspense } from 'react';
```
Comment (aadamw) [inline apps/web/src/app/[locale]/tracks/_components/track-enrolled-section-carousel.tsx:6]: Not used

Comment (Nkemjikanma) [reply apps/web/src/app/[locale]/tracks/_components/track-enrolled-section-carousel.tsx]: Haunted by unused imports 😢

**Review by aadamw — APPROVED** [review]

**Review by bautistaaa — APPROVED** [review]

**Pull request merged**